### ISSUE TITLE
Connection name list

### ISSUE THREAD

**Issue by deliego-github** [dsuch/pymqi#153]
Hi guys,

The way that PyMQI connects, does it support connection name lists?
We have a user with a multi-instance queue manager across servers and if the first host in the connection name list fails to connect, it does not attempt to connect to the 2nd or 3rd in the list.
We have tried modifying the mqclient.ini with parameter `DefRecon=YES` and also adding MQCNO_RECONNECT to the connection properties of the pymqi connection and still fails on the first host if unreachable.

From a comment on this post which I think was JoshMQ responding, it seems that it's supposed to work:
https://stackoverflow.com/questions/42312860/can-i-use-pymqi-to-connect-to-an-ibm-mq-multi-instance-queue-manager

My code is:
```
cd = pymqi.CD()
cd.ChannelName = 'SERVER_CONN_CHANNEL'
cd.ConnectionName = '192.168.180.2(1415),192.168.180.3(1415)'
cd.ChannelType = MQCHT_CLNTCONN
cd.TransportType = MQXPT_TCP
        
connect_options = MQCNO_HANDLE_SHARE_BLOCK|MQCNO_RECONNECT

qmgr = pymqi.QueueManager(None)
qmgr.connect_with_options(queueManager, cd, sco, user=username, password=password, opts=connect_options)
```

Any reason why it would not attempt to connect to the rest of the hosts in the list?

Thanks!

**Comment by JoshMcMQ** [comment]
What version of MQ libraries do you have?

**Comment by chughts** [comment]
@JoshMcMQ makes a good point. You need to ensure that the cd you construct conforms to a version that supports connection strings. eg. 

```python
            cd = pymqi.CD(Version=pymqi.CMQXC.MQCD_VERSION_11)
```

**Comment by deliego-github** [comment]
MQ Client: 9.01
pymqi: 1.9.2

So the CD structure needs to be version 11?

**Comment by SeyfSV** [comment]
Hi @deliego-github ,
I checked provided code on release version  - reconnect works as designed.
What error did you receive?

**Comment by deliego-github** [comment]
I can try using a newer build and adding the MQCD_VERSION_11 parameter.
The error returned is FAILED: MQRC_HOST_NOT_AVAILABLE

**Comment by SeyfSV** [comment]
`MQCD_VERSION_11` is overhead, `ConnectionName` parameter is part of `MQCD_VERSION_2` and default version is `MQCD_VERSION_6`.
Reconnect behaviour realised in MQ Client not in pymqi.
Do you connect to this Queue Managers if you use only:
`cd.ConnectionName = '192.168.180.3(1415)'`
Do you use python2.7?

**Comment by deliego-github** [comment]
Python 3.6
Let me confirm a few settings. I think it could be some MQ Client or server setting. I will let you know.

**Comment by deliego-github** [comment]
After a couple of tweaks, it seems to work but takes a bit of time to continue in the connection list.
At the end user, it takes almost 2 minutes to connect to the 2nd IP in the connection list, and this causes my application to throw an execution timeout.
So I believe this is something that could possibly be tweaked on the MQ Client perhaps... to force a connection timeout, perhaps in the TCP stanza
https://www.ibm.com/support/knowledgecenter/SSFKSJ_9.1.0/com.ibm.mq.con.doc/q016910_.htm

Since this is a cpython class, I am hoping this Connect_Timeout value may do the trick

I'll let you know the results once I hear back from the enduser

**Comment by JoshMcMQ** [comment]
Without `Connect_Timeout` set MQ times out based on OS defaults for unreachable hosts.   RHEL Linux defaults to ~186 seconds.   If the host was up but the channel stopped or listener down then the second IP world be tried almost immediately.

**Comment by chughts** [comment]
How are you testing? ie. How are you making the 1st server in the connection string unavailable?

**Comment by SeyfSV** [comment]
Who are you asking @chughts?
For example, I use 127.0.0.1(22)  as 1st server: MQ can establish TCP connection, but can't setup MQ connection...

**Comment by chughts** [comment]
I was asking @deliego-github . Reason or asking is that I had a similar problem when using MQ images running in docker, but this was because I was pausing the 'stopped' MQ servers. When I actually stopped the docker images the fail-over was 'instant'. 

Leading me to think that by just pausing the image, actually kept open some sort of network linkage open that required the normal network connection timeout of approx 3 mins to kick in.

**Comment by deliego-github** [comment]
That was exactly it. Connect_Timeout value in the mqclient.ini solved it and no more problems. It was using the OS default as Josh suggested, which it would explain why I was getting different results on different virtual instances I had (Windows was much quicker). @chughts, I tested by simply using a made up IP as the first entry in the CD. It was failing on my side because my app was timing out before it even got to test the second connection. Once I changed that I saw that it kept going but almost 3 minutes later which proved that it works, but it's just delayed so it lead me to the connection timeout parameter.

This is great, thanks for all your help guys!
